stop using CRD v1beta1 in tests

deads2k opened the pull request:
Remove unnecessary v1beta1 dependencies from tests.  This has a few we cannot yet remove, so I have to tweak those tests so they still test something valuable.  I'll remove WIP once I have done that.

/kind cleanup
/priority important-soon
@kubernetes/sig-api-machinery-pr-reviews 

```release-note
NONE
```

deads2k committed:
try to tidy up crd v1beta1  usage

deads2k committed:
attempt to coerce test for non-structural schemas into something close to valid

deads2k committed:
coerce apply tests to CRD v1. skip CRDs that are no longer valid

deads2k committed:
switch apiserver integration tests to CRD v1

deads2k commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/validation_test.go:
```diff
@@ -664,58 +638,64 @@ func TestForbiddenFieldsInSchema(t *testing.T) {
 }
 
 func TestNonStructuralSchemaConditionUpdate(t *testing.T) {
+	t.Skip("non-structural schemas can no longer be created")
```
This test has to be restored and use the v1beta1 client.

deads2k commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/validation_test.go:
```diff
@@ -862,6 +842,7 @@ spec:
 }
 
 func TestNonStructuralSchemaCondition(t *testing.T) {
+	t.Skip("non-structural schemas can no longer be created")
```
This test has to be restored and use the v1beta1 client.

deads2k commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/validation_test.go:
```diff
@@ -664,58 +638,64 @@ func TestForbiddenFieldsInSchema(t *testing.T) {
-			}
-
-			validationSchema.OpenAPIV3Schema.Properties["zeta"] = apiextensionsv1beta1.JSONSchemaProps{
+			// v1 doesn't allow additional properties at all. this isn't necessary in v1.
```
This needs to be put back to v1beta1 until 1.22, when it can be removed because it is testing client behavior, not etcd/controller behavior.

sttts replied:
we support it, but not at the root.

sttts replied:
as this to the schema instead:
```
validationSchema.OpenAPIV3Schema.Properties["eta"]: {
  type: object
  additionalProperties: false
}
```

deads2k commented on test/integration/apiserver/apply/apply_crd_beta_test.go:
```diff
@@ -0,0 +1,234 @@
+
+// TestApplyCRDNoSchema tests that CRDs and CRs can both be applied to with a PATCH request with the apply content type
+// when there is no validation field provided.
+func TestApplyCRDNoSchema(t *testing.T) {
```
This is a straight move.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/basic_test.go:
```diff
@@ -25,7 +25,8 @@ import (
 
-	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+
```
nit: no empty line

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion/conversion_test.go:
```diff
@@ -335,7 +336,7 @@ func validateNonTrivialConverted(t *testing.T, ctc *conversionTestContext) {
 
 			fixture := newConversionMultiVersionFixture(ns, name, createVersion.Name)
-			if !*ctc.crd.Spec.PreserveUnknownFields {
+			if createVersion.Schema.OpenAPIV3Schema.XPreserveUnknownFields == nil || !*createVersion.Schema.OpenAPIV3Schema.XPreserveUnknownFields {
```
not sure about this change. Do we have fixtures that make use of this schema field? Note that we also still have `crd.Spec.PreserveUnknownFields` in v1, settable through v1beta1 conversion.

sttts replied:
We don't. Check `multiVersionFixture`: it is static and has `PreserveUnknownFields` set to false, i.e. the if always applies. Just do the same here: remove the if and leave the `SetNestedField` with `foo: garbage`.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/limit_test.go:
```diff
@@ -22,9 +22,10 @@ import (
 	"testing"
 
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+
```
order

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/listtype_test.go:
```diff
@@ -22,37 +22,38 @@ import (
 	"testing"
 	"time"
 
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
```
order

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/pruning_test.go:
```diff
@@ -41,27 +43,37 @@ import (
+				},
+				Schema: &apiextensionsv1.CustomResourceValidation{
+					OpenAPIV3Schema: &apiextensionsv1.JSONSchemaProps{
+						XPreserveUnknownFields: pointer.BoolPtr(false),
```
no need to set this to false

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/scope_test.go:
```diff
@@ -21,6 +21,8 @@ import (
 	"testing"
 
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+
```
order

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/subresources_test.go:
```diff
@@ -424,21 +389,16 @@ func TestValidationSchemaWithStatus(t *testing.T) {
-	_, err = fixtures.CreateNewCustomResourceDefinition(noxuDefinition, apiExtensionClient, dynamicClient)
-	if err == nil {
-		t.Fatalf(`unexpected non-error, expected: must not have "additionalProperties" at the root of the schema if the status subresource is enabled`)
-	}
```
why is this removed?

deads2k replied:
> why is this removed?

a side effect of trying to make the schema eventually creatable.  Some tests relied on having a single schema for all versions.  Others wanted two versions.

sttts replied:
It amazes me that we have an invalid schema in a standard fixture.

sttts replied:
ok, found it: for structural schemas we have this:

```
	if g.AdditionalProperties != nil {
		if lvl == rootLevel {
			allErrs = append(allErrs, field.Forbidden(fldPath.Child("additionalProperties"), "must not be used at the root"))
		}
```

In other words, we forbid `additionalProperties` even for CRDs without status subresource. So removing this extra test is fine. Please also remove the now invalid comment above the removal.

fedebongio commented:
/triage accepted

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/apiapproval_test.go:
```diff
@@ -21,6 +21,8 @@ import (
 	"time"
 
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+
```
nit: empty line

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion/conversion_test.go:
```diff
@@ -27,6 +27,8 @@ import (
 	"testing"
 	"time"
 
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
```
order

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion/conversion_test.go:
```diff
@@ -393,7 +394,7 @@ func validateNonTrivialConvertedList(t *testing.T, ctc *conversionTestContext) {
 		client := ctc.versionedClient(ns, createVersion.Name)
 		fixture := newConversionMultiVersionFixture(ns, name, createVersion.Name)
-		if !*ctc.crd.Spec.PreserveUnknownFields {
+		if createVersion.Schema.OpenAPIV3Schema.XPreserveUnknownFields == nil || !*createVersion.Schema.OpenAPIV3Schema.XPreserveUnknownFields {
```
same here, remove the whole check and lieve the `SetNestedField` (unconditionally)

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion/conversion_test.go:
```diff
@@ -428,13 +429,12 @@ func validateNonTrivialConvertedList(t *testing.T, ctc *conversionTestContext) {
 	for _, createVersion := range ctc.crd.Spec.Versions {
+		if createVersion.Schema.OpenAPIV3Schema.XPreserveUnknownFields == nil || !*createVersion.Schema.OpenAPIV3Schema.XPreserveUnknownFields {
+			continue
+		}
```
also here. I don't think this is used at all.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go:
```diff
@@ -22,9 +22,7 @@ import (
-
-	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
-
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
```
order

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/validation_test.go:
```diff
@@ -123,81 +120,69 @@ func newNoxuValidationCRDs(scope apiextensionsv1beta1.ResourceScope) []*apiexten
-						Minimum: float64Ptr(0),
-					},
-				},
-			},
```
Impossible to follow these removals.

sttts replied:
I remember you asked me on slack. Yes, this is not structural. If you have removed the delta test code as well, I think we are fine.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/subresources_test.go:
```diff
@@ -489,24 +454,18 @@ func TestValidateOnlyStatus(t *testing.T) {
+			OpenAPIV3Schema: schema,
+		}
+		schemaWithDescription := schema.DeepCopy()
+		schemaWithDescription.Description = "test"
```
I have no idea why we had set the decription at all. Would completely remove that `schemaWithDescription` code here and see whether it's green.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/validation_test.go:
```diff
@@ -530,18 +515,6 @@ func TestCustomResourceValidationErrors(t *testing.T) {
-					"must validate at least one schema (anyOf)",
-					"delta in body should be at most 5 chars long",
-				},
-			},
```
ack, here it is.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/validation_test.go:
```diff
@@ -664,58 +638,64 @@ func TestForbiddenFieldsInSchema(t *testing.T) {
+			// v1 doesn't allow additional properties at all. this isn't necessary in v1.
+			//validationSchema.OpenAPIV3Schema.AdditionalProperties.Allows = true
+
+			_, err = fixtures.CreateNewV1CustomResourceDefinition(noxuDefinition, apiExtensionClient, dynamicClient)
```
do this instead:
```
validationSchema.OpenAPIV3Schema.Properties["eta"]: {
  type: object
  additionalProperties:
    type: string
}
```

sttts commented on test/integration/apiserver/apiserver_test.go:
```diff
@@ -32,9 +32,10 @@ import (
 	"testing"
 	"time"
 
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
```
order

deads2k committed:
fix up CRD validation tests

deads2k committed:
switch condition controller check to use a direct etcd write

deads2k committed:
re-enable CRD v1beta1 validation test

deads2k commented:
@sttts pushed commits addressing your comments.  Brought back the two v1beta1 validation tests.  One I wired directly against etcd. The second one I've left as an exercise for you to unwind.  I suggest splitting into "things still valid in v1" and "the rest". Then we have to remove "the rest" when we branch master.

deads2k committed:
generated

deads2k committed:
write CRD validation function only valid in beta1 direct to etcd to test apply on migrated data

sttts commented:
/lgtm

sttts commented:
/lgtm

deads2k commented:
/retest

sttts commented:
/retest

sttts commented:
/lgtm
